### PR TITLE
python3Packages.spacy: Add dependencies for 3.0.5

### PR DESCRIPTION
**Pull request by mattmelling**
###### Motivation for this change
https://nix-cache.s3.amazonaws.com/log/a19dxg5y85pl6i7lqkiayfg5nam2bs2s-python3.8-spacy-3.0.5.drv

Adds dependencies required for spaCy 3.0.5.

ZHF: #122042

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
  - [X] NixOS
  - [ ] macOS
  - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/pathy/default.nix [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-nb8my/5rkc7thuHnXZHe1Hg8j+sLBlYyJcLHWrrKZ5M=";
+  };
+
+  propagatedBuildInputs = [ smart_open typer google-cloud-storage ];
+
+  postPatch = ''
+    substituteInPlace requirements.txt --replace "smart-open>=2.2.0,<4.0.0" "smart-open>=2.2.0"
```
Comment (jonringer) [inline pkgs/development/python-modules/pathy/default.nix:24]: just a little more readable
```suggestion
    substituteInPlace requirements.txt \
      --replace "smart-open>=2.2.0,<4.0.0" "smart-open>=2.2.0"
```

File: pkgs/development/python-modules/pathy/default.nix [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  '';
+
+  checkInputs = [ pytest mock ];
```
Comment (jonringer) [inline pkgs/development/python-modules/pathy/default.nix:27]: please use `pytestCheckHook` on new packages

depending on existing `setup.py test` has been deprecated by setuptools

File: pkgs/development/python-modules/spacy/legacy.nix [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  pname = "spacy-legacy";
+  version = "3.0.5";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-Uy94rjFllSj622RTzd6UJaQmIniCw4gpeq/X57QcIpA=";
+  };
+
+  # checkInputs = [ pytest spacy ];
+  doCheck = false;
```
Comment (jonringer) [inline pkgs/development/python-modules/spacy/legacy.nix:16]: There are no tests included. Please try to checkout from source and check if they have unit tests, and try to run them. Unit tests give a good indication that they package has a high degree of validity and correctness given the python package set.

If tests are not available, then please use `pythonImportsCheck` to import the most important modules. This isn't as good as unit tests, but will usually give a good indication of run-time errors. Please see [pythonImportsCheck documentation](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pythonimportscheck) for more information.

File: pkgs/top-level/python-packages.nix [hunk 4]
```diff
@@ -7611,6 +7613,8 @@ in {
 
   spacy = callPackage ../development/python-modules/spacy { };
 
+  spacy_legacy = callPackage ../development/python-modules/spacy/legacy.nix { };
```
Comment (jonringer) [inline pkgs/top-level/python-packages.nix:7616]: Please normalize the attribute name. According to the [python contributing guidelines](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#contributing-guidelines), package names within nix should follow [PEP 0503](https://www.python.org/dev/peps/pep-0503/#normalized-names)

`spacy-legacy`

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 122112` at f29655d4 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>python38Packages.textacy</li>
  </ul>
</details>
<details>
  <summary>4 packages built successfully:</summary>
  <ul>
    <li>python38Packages.pathy</li>
    <li>python38Packages.spacy</li>
    <li>python38Packages.spacy_legacy</li>
    <li>python39Packages.spacy_legacy</li>
  </ul>
</details>
<details>
  <summary>7 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/stale-substitute.md">stale-substitute</a>

Stale substituteInPlace detected.
substituteStream(): WARNING: pattern 'catalogue>=0.0.7,<1.1.0' doesn't match anything in file 'setup.cfg'
Near pkgs/development/python-modules/spacy/default.nix:82:0:
```
   |
82 |     description = "Industrial-strength Natural Language Processing (NLP) with Python and Cython";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md">python-include-tests</a>

Consider adding a `checkPhase` for tests, or if not feasible, `pythonImportsCheck`.

Near pkgs/development/python-modules/spacy/legacy.nix:19:0:
```
   |
19 |     description = "A Path interface for local and cloud bucket storage";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-python-tests.md">no-python-tests</a>

Test runner could not discover any test cases: ‘Ran 0 tests in 0.000s’
Near pkgs/development/python-modules/pathy/default.nix:30:0:
```
   |
30 |     description = "A Path interface for local and cloud bucket storage";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/stale-substitute.md">stale-substitute</a>

Stale substituteInPlace detected.
substituteStream(): WARNING: pattern 'plac>=0.9.6,<1.2.0' doesn't match anything in file 'setup.cfg'
Near pkgs/development/python-modules/spacy/default.nix:82:0:
```
   |
82 |     description = "Industrial-strength Natural Language Processing (NLP) with Python and Cython";
   | ^
```
</li>
    <li>warning: unused-argument

Unused argument: `dataclasses`.
Near pkgs/development/python-modules/pathy/default.nix:5:3:
```
  |
5 | , dataclasses
  |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/stale-substitute.md">stale-substitute</a>

Stale substituteInPlace detected.
substituteStream(): WARNING: pattern 'thinc>=7.4.1,<7.5.0' doesn't match anything in file 'setup.cfg'
Near pkgs/development/python-modules/spacy/default.nix:82:0:
```
   |
82 |     description = "Industrial-strength Natural Language Processing (NLP) with Python and Cython";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/stale-substitute.md">stale-substitute</a>

Stale substituteInPlace detected.
substituteStream(): WARNING: pattern 'srsly>=1.0.2,<1.1.0' doesn't match anything in file 'setup.cfg'
Near pkgs/development/python-modules/spacy/default.nix:82:0:
```
   |
82 |     description = "Industrial-strength Natural Language Processing (NLP) with Python and Cython";
   | ^
```
</li>
  </ul>
</details>

File: pkgs/development/python-modules/spacy/legacy.nix (hunk 3 above)
Comment (mattmelling) [reply pkgs/development/python-modules/spacy/legacy.nix]: Have added `pythonImportsCheck`, this package's tests seem to depend on `spacy` which creates a circular dependency.

**Review by jonringer — APPROVED** [review]
LGTM

Result of nixpkgs-review pr 122112 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)<details>  <summary>1 package marked as broken and skipped:</summary>  <ul>    <li>python38Packages.textacy</li>  </ul></details><details>  <summary>4 packages built:</summary>  <ul>    <li>python38Packages.pathy</li>    <li>python38Packages.spacy</li>    <li>python38Packages.spacy-legacy</li>    <li>python39Packages.spacy-legacy</li>  </ul></details>

**Pull request merged**